syntax = "proto3";

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc ClientRequest(ClientRequestMessage) returns (ClientResponseMessage);
    rpc GetState(Empty) returns (StateResponse);
}

message Empty {}

message RequestVoteRequest {
    int32 term = 1;
    int32 candidateId = 2;
}

message RequestVoteResponse {
    int32 term = 1;
    bool voteGranted = 2;
}

message AppendEntriesRequest {
    int32 term = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    int32 leaderCommit = 6;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message LogEntry {
    int32 term = 1;
    int32 index = 2;
    string operation = 3;
}

message ClientRequestMessage {
    string operation = 1;
}

message ClientResponseMessage {
    string result = 1;
}

message StateResponse {
    int32 id = 1;
    int32 term = 2;
    string state = 3;
    int32 commitIndex = 4;
    int32 lastApplied = 5;
    repeated LogEntry log = 6;
}
